### ISSUE TITLE
[BUG/ISSUE] Apparent parallelization error in KPP, probably in gckpp_Hetrates.F90

### ISSUE THREAD

**Issue by yantosca** [geoschem/geos-chem#392]
Both @cdholmes and I have observed what may be a parallelization error in the KPP code, probably in the gckpp_HetRates.F90 module.

When I was doing difference tests to validate PR https://github.com/geoschem/geos-chem/pull/386, I noticed that when I turned OpenMP parallelization off, I got zero differences in the SpeciesConc collection:
```
Variable               Ref=refrundirsp          Dev=devrundirsp          Dev-Ref
AREA                 : 510705170000000.0      | 510705170000000.0      | 0.0 
SpeciesConc_A3O2     : 3.693731e-10           | 3.693731e-10           | 0.0 
SpeciesConc_ACET     : 0.00014893242          | 0.00014893242          | 0.0 
SpeciesConc_ACTA     : 1.6614892e-07          | 1.6614892e-07          | 0.0 
SpeciesConc_AERI     : 8.852037e-08           | 8.852037e-08           | 0.0 
SpeciesConc_ALD2     : 1.1507503e-05          | 1.1507503e-05          | 0.0 
SpeciesConc_ALK4     : 1.5890859e-05          | 1.5890859e-05          | 0.0 
SpeciesConc_ASOA1    : 3.3509856e-08          | 3.3509856e-08          | 0.0 
SpeciesConc_ASOA2    : 1.3462016e-08          | 1.3462016e-08          | 0.0 
SpeciesConc_ASOA3    : 2.2050894e-08          | 2.2050894e-08          | 0.0 
SpeciesConc_ASOAN    : 2.3715214e-07          | 2.3715214e-07          | 0.0 
SpeciesConc_ASOG1    : 3.255437e-08           | 3.255437e-08           | 0.0 
SpeciesConc_ASOG2    : 4.4420226e-08          | 4.4420226e-08          | 0.0 
SpeciesConc_ASOG3    : 4.648521e-07           | 4.648521e-07           | 0.0 
SpeciesConc_ATO2     : 2.3170594e-09          | 2.3170594e-09          | 0.0 
SpeciesConc_ATOOH    : 1.0495365e-06          | 1.0495365e-06          | 0.0 
SpeciesConc_B3O2     : 1.3071982e-09          | 1.3071982e-09          | 0.0 
SpeciesConc_BCPI     : 6.134594e-06           | 6.134594e-06           | 0.0 
SpeciesConc_BCPO     : 1.2362641e-06          | 1.2362641e-06          | 0.0 
SpeciesConc_BENZ     : 5.4437505e-06          | 5.4437505e-06          | 0.0 
... etc ...
```
But when I had OpenMP parallelzation turned on, some species showed very small differences, of the order of numerical noise (i.e. 5-7 orders of magnitude lower than the values themselves):
```
Variable               Ref=refrundir            Dev=devrundir            Dev-Ref
AREA                 : 510705170000000.0      | 510705170000000.0      | 0.0 
SpeciesConc_A3O2     : 3.9885814e-10          | 3.9886247e-10          | 4.3298697960381105e-15 
SpeciesConc_ACET     : 0.00014896577          | 0.00014896577          | 0.0 
SpeciesConc_ACTA     : 1.7908036e-07          | 1.7908037e-07          | 1.4210854715202004e-14 
SpeciesConc_AERI     : 8.855729e-08           | 8.8557286e-08          | -7.105427357601002e-15 
SpeciesConc_ALD2     : 1.150178e-05           | 1.1501781e-05          | 9.094947017729282e-13 
SpeciesConc_ALK4     : 1.5931619e-05          | 1.5931619e-05          | 0.0 
SpeciesConc_ASOA1    : 3.3577823e-08          | 3.357782e-08           | -3.552713678800501e-15 
SpeciesConc_ASOA2    : 1.3498284e-08          | 1.34982825e-08         | -1.7763568394002505e-15 
SpeciesConc_ASOA3    : 2.2045986e-08          | 2.2045986e-08          | 0.0 
SpeciesConc_ASOAN    : 2.3719303e-07          | 2.3719303e-07          | 0.0 
SpeciesConc_ASOG1    : 3.2562205e-08          | 3.2562205e-08          | 0.0 
SpeciesConc_ASOG2    : 4.4528498e-08          | 4.4528498e-08          | 0.0 
SpeciesConc_ASOG3    : 4.6504647e-07          | 4.6504647e-07          | 0.0 
SpeciesConc_ATO2     : 2.34998e-09            | 2.3499815e-09          | 1.5543122344752192e-15 
SpeciesConc_ATOOH    : 1.047004e-06           | 1.0470042e-06          | 1.1368683772161603e-13 
SpeciesConc_B3O2     : 1.4223676e-09          | 1.4223753e-09          | 7.771561172376096e-15 
SpeciesConc_BCPI     : 6.1385153e-06          | 6.1385153e-06          | 0.0 
SpeciesConc_BCPO     : 1.2475313e-06          | 1.2475313e-06          | 0.0 
SpeciesConc_BENZ     : 5.473232e-06           | 5.473232e-06           | 0.0 
...etc...
```
@cdholmes had noticed a similar result.  When he introduced a structural update (building off 12.9.2) that should not have resulted in differences, small numerical differences of similar magnitude were observed.

I suspect that this is caused by a variable that is not being held PRIVATE or THREADPRIVATE for the OpenMP parallel loop.

**Comment by yantosca** [comment]
I have run 1-hr difference tests:

- Ref: HetChem on, OpenMP off
- Dev: HetChem on, OpenMP on

and

- Ref: HetChem off, OpenMP off
- Dev: HetChem off, OpenMP on

and have obtained identical results each time.  Still not sure where numerical differences are occuring.  It could be due to optimization.  I will keep an eye on this.

**Comment by cdholmes** [comment]
I get a different result. I do see differences with OpenMP on vs. off. I'm running 12.9.2 release version, tropchem mechanism, 4x5, MERRA2, compiled with ifort 19.0.5.281. Chemistry is on but all other processes are off (emissions, transport, mixing, dry/wet deposition). Each simulation is 20 minutes. The only change between the simulations is 
make OMP=y build
make OMP=n build

**Comment by yantosca** [comment]
I think it is still very possible that there still is a parallelization error.  Maybe my difftest by luck just ended up parallelizing Dev and Ref in the same way.

**Comment by yantosca** [comment]
Also -- you can try setting DO_HETCHEM=.FALSE. in flexchem_mod.F90 and try a difftest on that.  If you get identical results then you know the error is in the hetrates.  I will try a similar test again.

**Comment by cdholmes** [comment]
After repeating my parallelization test with the Unit Tester, I find /no/ difference between runs with OpenMP on and off. This is with HetChem on. 

I still find, however, that purely structural changes to gckpp_HetRates.F90 (eg replacing string comparison with integer comparison having same logical structure), are causing  differences between runs when none are expected.

**Comment by yantosca** [comment]
I will remove this issue from the 13.0.0 milestone, as it seems we do not have a definite resolution.  I will leave the issue open however.

**Comment by yantosca** [comment]
This issue is now fixed by commit 1291ab2a91eb4e76e7d08837faeeb7d2e18eef71, which for the moment is currently in the branch **bugfix/bmy/SulfateModOmpFix**.  This will be merged into the final 13.0.0 version.

Although it initially seemed that this error was in `gckpp_HetRates.F90`, further investigation revealed that it was actually in the CHEM_SO2 routine of `sulfate_mod.F90`.  Pointers to `State_Chm%isCloud`, `State_Chm%pHCloud`, and `State_Chm%QLxpHCloud` were added to the parallel loop but never declared as `!$OMP PRIVATE`.  Furthermore, the `TDCA` variable was also not declared `!$OMP PRIVATE`.  This commit resolves all of these issues.

The error propagated to the KPP mechanism via the heterogeneous chemistry rates.  Subsequent simulations revealed slightly different results due to the values of pHCloud not being computed identically at each (I,J,L).  This affected the rates of species like HBr, HOBr, and HCl.
